File: FieldMark/FieldMark.Domain/ValueObjects/Role.cs
namespace FieldMark.Domain.ValueObjects;

/// <summary>
/// Conceptual role of an authenticated FieldMark user. The five canonical
/// names are persisted in dotnet_auth, django_auth, and fiber_auth — this
/// type is the single .NET-side source of truth for them.
/// </summary>
public sealed record Role
{
    public string Name { get; }
    public string Label { get; }
    public string BadgeToken { get; }

    private Role(string name, string label, string badgeToken)
    {
        Name = name;
        Label = label;
        BadgeToken = badgeToken;
    }

    public static readonly Role Admin = new("ADMIN", "Admin", "danger");
    public static readonly Role ComplianceOfficer = new(
        "COMPLIANCE_OFFICER",
        "Compliance Officer",
        "info"
    );
    public static readonly Role Inspector = new("INSPECTOR", "Inspector", "warning");
    public static readonly Role SiteSupervisor = new(
        "SITE_SUPERVISOR",
        "Site Supervisor",
        "neutral"
    );
    public static readonly Role Executive = new("EXECUTIVE", "Executive", "success");

    public static IReadOnlyList<Role> All { get; } =
        new[] { Admin, ComplianceOfficer, Inspector, SiteSupervisor, Executive };

    public static Role Parse(string name) =>
        All.FirstOrDefault(r => r.Name == name)
        ?? throw new ArgumentException($"Unknown role name: {name}", nameof(name));

    public override string ToString() => Name;
}

File: fieldmark-go/internal/domain/role.go
package domain

// Role is the single Go-side source of truth for the five canonical
// conceptual-role names; mirrored in dotnet_auth, django_auth, and
// fiber_auth.user_roles.role (CHECK constraint).
type Role string

const (
	RoleAdmin             Role = "ADMIN"
	RoleComplianceOfficer Role = "COMPLIANCE_OFFICER"
	RoleInspector         Role = "INSPECTOR"
	RoleSiteSupervisor    Role = "SITE_SUPERVISOR"
	RoleExecutive         Role = "EXECUTIVE"
)

// AllRoles enumerates the canonical names in deterministic order.
var AllRoles = []Role{
	RoleAdmin, RoleComplianceOfficer, RoleInspector, RoleSiteSupervisor, RoleExecutive,
}

// Label returns the title-cased display label for the role (AC #4, Story 1.13).
func (r Role) Label() string {
	switch r {
	case RoleAdmin:
		return "Admin"
	case RoleComplianceOfficer:
		return "Compliance Officer"
	case RoleInspector:
		return "Inspector"
	case RoleSiteSupervisor:
		return "Site Supervisor"
	case RoleExecutive:
		return "Executive"
	default:
		return ""
	}
}

// BadgeToken returns the CSS badge modifier token for the role (AC #4, Story 1.13).
func (r Role) BadgeToken() string {
	switch r {
	case RoleAdmin:
		return "danger"
	case RoleComplianceOfficer:
		return "info"
	case RoleInspector:
		return "warning"
	case RoleSiteSupervisor:
		return "neutral"
	case RoleExecutive:
		return "success"
	default:
		return "neutral"
	}
}
